### DISCUSSION TITLE
Using noodles_tabix for reading tbi-indexed tsv.bgz files

### DISCUSSION THREAD

**Discussion by holtgrewe** [zaeleus/noodles#170] [Q&A] [answered]
Hi @zaeleus - could you give an example how I could use the `noodles_tabix` module for jumping in bgzip-ed and tabix indexed TSV files?

**Comment by zaeleus** [comment, marked as answer]
The general strategy to querying is to

  1) find the relevant chunks in the index ([`csi::Index::query`](https://docs.rs/noodles/0.37.0/noodles/csi/struct.Index.html#method.query)),
  2) read the data between those chunks ([`csi::io::Query`](https://docs.rs/noodles/latest/noodles/csi/io/struct.Query.html)),
  3) parse the records within that data, and
  4) filter the records that intersect the region.

There is currently no way to query an arbitrary tabix-indexed file, but it is planned as part of #158. You can, however, implement 3 and 4. I added an example of this in [`noodles-tabix/examples/tabix_query_generic.rs`](https://github.com/zaeleus/noodles/blob/a7eb478bc19dc5de6cbd0392e07cbaa0ec23f00b/noodles-tabix/examples/tabix_query_generic.rs).

**Comment by zaeleus** [comment]
noodles 0.39.0/noodles-csi 0.18.0 now supports reading generic indexed files. It introduces a record parser ([`IndexedRecords`](https://docs.rs/noodles/0.39.0/noodles/csi/io/struct.IndexedRecords.html)) and filter ([`FilterByRegion`](https://docs.rs/noodles/0.39.0/noodles/csi/io/struct.FilterByRegion.html)), e.g.,

```rust
let records = Query::new(&mut decoder, chunks)
    .indexed_records(&header)
    .filter_by_region(&region);

for result in records {
    let record = result?;
    // ...
}
```

This pattern is wrapped by [`tabix::io::IndexedReader`](https://docs.rs/noodles/0.39.0/noodles/csi/io/struct.IndexedReader.html), which simplifies [`noodles-tabix/examples/tabix_query_generic.rs`](https://github.com/zaeleus/noodles/blob/noodles-0.39.0/noodles-tabix/examples/tabix_query_generic.rs).

**Reply by brentp** [reply]
Hi @zaeleus , I'm trying to further understand this. 
Is there an abstraction / trait that unifies reading from an indexed BCF or VCF where the index is either CSI or TBI?

**Reply by zaeleus** [reply]
Not currently, but I'll add one! Tracked in #188.
